### DISCUSSION TITLE
Debug session for two different languages at the same time

### DISCUSSION THREAD

**Discussion by ptrthomas** [microsoft/vscode-discussions#1052] [Extension Development QnA] [answered]
I have written an extension to support a custom language and also implemented a [DAP server](https://microsoft.github.io/debug-adapter-protocol) and it works fine in VS Code. So far so good.

It so happens that the custom language can invoke Java code. So now I want to enhance the debug capability so that in the same debug session you can stop at both custom-language *and* Java breakpoints.

Is this at all possible in VS Code? Are there examples of plugins that do this? I guess one way to approach this is to implement a [debugger extension](https://code.visualstudio.com/api/extension-guides/debugger-extension) in Typescript that a) is able to switch "modes" and b) delegate to two different DAP servers (running on different ports).

Any advice or references would be greatly appreciated!

**Comment by theReynald** [comment, marked as answer]
Hi @ptrthomas .

Enhancing debug capabilities in VS Code to support both custom language and Java breakpoints within the same debug session is indeed possible. Here are some approaches and considerations:

1. **Implement Multi-language Debug Adapter**: As you suggested, you can create a single debug adapter extension that supports both your custom language and Java. This debug adapter would need to handle switching between "modes" and delegate debugging requests to the appropriate DAP servers for each language. You would need to implement logic to handle breakpoints, stepping, variable inspection, etc., for both languages.

2. **Use Multi-target Debuggers**: Some debug adapters support debugging multiple languages simultaneously by allowing you to attach multiple debug configurations or targets to a single debug session. You could explore whether your custom language and Java debugger extensions support this feature. If they do, you can configure both debuggers to run in parallel within the same session.

3. **Extend Existing Debuggers**: If there are existing debugger extensions for your custom language and Java, you could extend them to support the other language. This approach may require modifying the existing debuggers' codebases to add support for the additional language and integrating them into a single extension.

4. **Develop Custom Debugging Workflow**: Depending on the complexity and requirements of your project, you may need to develop a custom debugging workflow that integrates separate debuggers for your custom language and Java. This could involve running multiple instances of VS Code or using VS Code's multi-root workspace feature to debug different parts of your project separately.

Unfortunately, I'm not aware of existing examples of plugins that precisely match your requirements, as this scenario can be highly specific to your project's architecture and language requirements. However, you can leverage the VS Code documentation, sample extensions, and community forums for guidance and assistance as you develop your custom debugger extension.

Remember to thoroughly test your solution to ensure that debugging both languages within the same session works seamlessly and efficiently. Additionally, consider the user experience and provide clear documentation and instructions for configuring and using the multi-language debugging feature.

Thoughts?

**Reply by ptrthomas** [reply]
Thanks for the details ! for the benefit of anyone coming across this thread in future, option (2) worked for me. VS Code happens to nicely support multiple debug configurations live at the same time and running in parallel. When a breakpoint is hit, VS Code knows which session and console to switch focus to on the UI.
